File: java/src/test/java/org/hello/HelloTests.java
package org.hello;

import org.junit.jupiter.api.Test;
import org.mockito.Mockito;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

class HelloTests {

  @Test
  void shouldReturnHelloWorld() {
    var message = new HelloMessage();
    assertThat(message.getText()).isEqualTo("Hello World!");
  }

  @Test
  void shouldPrintHelloMessage() {
    var messageText = "Hello Test!";
    var message = mock(HelloMessage.class);
    when(message.getText()).thenReturn(messageText);

    var console = Mockito.mock(HelloConsole.class);

    var app = new HelloApp(message, console);
    app.printHello();

    verify(console).print(messageText);
  }
}

File: kotlin/src/test/kotlin/org/hello/HelloTests.kt
package org.hello

import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.shouldBe
import io.mockk.Runs
import io.mockk.every
import io.mockk.just
import io.mockk.mockk
import io.mockk.verify

class HelloTests : DescribeSpec({

  describe("HelloMessage") {
    it("should return hello world") {
      val message = HelloMessage()
      message.text shouldBe "Hello World!"
    }
  }

  describe("HelloApp") {
    it("should print hello message") {
      val messageText = "Hello Test!"
      val message = mockk<HelloMessage>()
      every { message.text } returns (messageText)

      val console = mockk<HelloConsole>()
      every { console.print(any()) } just Runs

      val app = HelloApp(message, console)
      app.printHello()

      verify { console.print(messageText) }
    }
  }
})
